ml4gw/transforms/decimator.py

```diff
@@ -0,0 +1,137 @@
+import torch
+
+
+class Decimator(torch.nn.Module):
+    """
+    Downsample/decimate timeseries according to a schedule.
+
+    The schedule is a tensor of shape (N, 3), where each row is:
+    [start_time, end_time, target_sample_rate]
+
+    Example schedule (seconds, Hz):
+    [[0, 40, 256],   # first 40s at 256 Hz
+     [40, 58, 512],  # next 18s at 512 Hz
+     [58, 60, 2048]] # last 2s at full 2048 Hz
+
+    Strain: Tensor of shape (B, C, T), where
+    B=batch size, C=channels, T=time.
+
+    Functions:
+        build_variable_indices(sample_rate, schedule, device)
+            Compute the sample indices to keep from the full strain.
+        split_by_schedule(strain, schedule)
+            Split a strain into segments according to the schedule.
+    """
+
+    def __init__(
+        self,
+        sample_rate: int = None,
+        schedule: torch.Tensor = None,
+    ):
+        super().__init__()
+        self.sample_rate = sample_rate
+        self.schedule = schedule
+
+        self._validate_inputs()
+        idx = self.build_variable_indices()
+        self.register_buffer("idx", idx)
+
+    def _validate_inputs(
+        self,
+        strain: torch.Tensor = None,
+    ):
+        """
+        Validate inputs for compatibility.
+        If `strain` is None, only validates schedule and sample_rate.
+        If `strain` is provided, validate only strain (called in forward).
+        """
+        if strain is None:
+            if self.schedule.ndim != 2 or self.schedule.shape[1] != 3:
+                raise ValueError(
+                    f"Schedule must be of shape (N, 3), got \
+                        {self.schedule.shape}"
+                )
+
+            if not torch.all(self.schedule[:, 1] > self.schedule[:, 0]):
+                raise ValueError(
+                    "Each schedule segment must have end_time > start_time"
+                )
+
+            if torch.any(self.sample_rate % self.schedule[:, 2].long() != 0):
+                raise ValueError(
+                    f"Sample rate {self.sample_rate} must be divisible by all "
+                    f"target rates {self.schedule[:, 2].tolist()}"
+                )
+
+        else:
+            if strain.ndim < 1:
+                raise ValueError(
+                    "strain must have at least 1 dimension (time axis)"
+                )
+
+            strain_len = int(
```

Comment:
This is something that could be calculated during `__init__`, and then in `forward` we could just check if `self.expected_size == strain.shape[-1]`. And then this validation function could be simplified so that it never needs to take a `strain` argument (I think it's fine to remove the check for strain dimensionality).